src/main/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdateService.java

```diff
@@ -21,124 +13,59 @@
+        return govUkPayEventService.storeEventAndUpdateStateForMandate(mandate, MANDATE_SUBMITTED);
     }
 
-    void changePaymentMethodFor(Mandate mandate) {
-        Mandate newMandate = updateStateFor(mandate, SupportedEvent.PAYMENT_CANCELLED_BY_USER_NOT_ELIGIBLE);
-        directDebitEventService.registerPaymentMethodChangedEventFor(newMandate);
-        govUkPayEventService.storeEventForMandate(mandate, MANDATE_CANCELLED_BY_USER_NOT_ELIGIBLE);
+    Mandate changePaymentMethodFor(Mandate mandate) {
+        return govUkPayEventService.storeEventAndUpdateStateForMandate(mandate, MANDATE_CANCELLED_BY_USER_NOT_ELIGIBLE);
     }
 
-    void cancelMandateCreation(Mandate mandate) {
-        Mandate newMandate = updateStateFor(mandate, SupportedEvent.PAYMENT_CANCELLED_BY_USER);
-        directDebitEventService.registerMandateCancelledEventFor(newMandate);
-        govUkPayEventService.storeEventForMandate(mandate, MANDATE_CANCELLED_BY_USER);
+    Mandate cancelMandateCreation(Mandate mandate) {
+        return govUkPayEventService.storeEventAndUpdateStateForMandate(mandate, MANDATE_CANCELLED_BY_USER);
     }
 
-    public DirectDebitEvent mandateActiveFor(Mandate mandate) {
-        return directDebitEventService.registerMandateActiveEventFor(mandate);
+    public Mandate mandateExpiredFor(Mandate mandate) {
+        return govUkPayEventService.storeEventAndUpdateStateForMandate(mandate, MANDATE_EXPIRED_BY_SYSTEM);
     }
 
-    public void mandateExpiredFor(Mandate mandate) {
-        Mandate updatedMandate = updateStateFor(mandate, SupportedEvent.MANDATE_EXPIRED_BY_SYSTEM);
-        directDebitEventService.registerMandateExpiredEventFor(updatedMandate);
-        govUkPayEventService.storeEventForMandate(mandate, MANDATE_EXPIRED_BY_SYSTEM);
+    Mandate tokenExchangedFor(Mandate mandate) {
+        return govUkPayEventService.storeEventAndUpdateStateForMandate(mandate, MANDATE_TOKEN_EXCHANGED);
     }
 
-    public DirectDebitEvent mandateFailedFor(Mandate mandate) {
-        userNotificationService.sendMandateFailedEmailFor(mandate);
-        return directDebitEventService.registerMandateFailedEventFor(mandate);
-    }
-
-    public DirectDebitEvent mandateCancelledFor(Mandate mandate) {
-        userNotificationService.sendMandateCancelledEmailFor(mandate);
-        return directDebitEventService.registerMandateCancelledEventFor(mandate);
-    }
-
-    public DirectDebitEvent mandatePendingFor(Mandate mandate) {
-        return directDebitEventService.registerMandatePendingEventFor(mandate);
-    }
-
-    public Mandate receiveDirectDebitDetailsFor(Mandate mandate) {
-        directDebitEventService.registerDirectDebitReceivedEventFor(mandate);
-        return mandate;
-    }
-
-    public void payerCreatedFor(Mandate mandate) {
-        directDebitEventService.registerPayerCreatedEventFor(mandate);
-    }
-
-    public void payerEditedFor(Mandate mandate) {
-        directDebitEventService.registerPayerEditedEventFor(mandate);
-    }
```

Comment:
Yeh but it isn't really anything we need to know about is it? It's just part of the user setup. The state of the mandate hasn't changed at all.

Reply:
I think in an ideal world, all changes (observable changes: changes to data, communication with external services, sending emails etc.) would be represented as events. To fully realise this vision, we’d need events to carry a payload allowing the events on their own to recreate the current state. We’re not there yet, so I can see that an event saying a payer has been edited (without any specifics) isn’t so useful.

Maybe in the future we can revisit this and — perhaps inspired by the work going on in ledger — make this more event-sourced. For now, the benefit is marginal (it could make tracing user journeys slightly easier), so I’m happy to let this go for now. The dream can wait.